### PR TITLE
Rebalanced Power consumption, increase for machines

### PR DESCRIPTION
**Pull request by Ghilker**
## About The Pull Request
This PR rebalances power consumption for all machines, increasing it and fixing many issues where machines wouldn't consume power at all.
Baseline is 100 W of power in idle and 1000 W in active state.
Many machines with parts now increase power consumption when upgraded, allowing the choice of faster production or less power consumed; this also is an incentive for engineering to try and improve the power generation in the late stage of the game 

Notice: The base SM with SMES properly configured is enough for the station when all machines are at tier 1. With few upgrades that don't touch the base cooling loop it can carry most of the rounds. 

[Feedback thread](https://tgstation13.org/phpBB/viewtopic.php?f=10&t=31490)

## Why It's Good For The Game
The increase in power consumption will push players to find a balance between upgrading parts and spending energy in the initial stages of the game, plus engies have now a reason to increment the power generated by the SM and improve the power storage with new smes

## Changelog

:cl:
balance: machines now consume more power all around. Baseline 100 W idle and 1000 W active
balance: machines with parts will consume more power depending on the part and on the part tier
balance: Notice: The base SM with SMES properly configured is enough for the station when all machines are at tier 1. With few upgrades that don't touch the base cooling loop it can carry most of the rounds. 
/:cl:

### CODE REVIEW

**Comment by bruhlookatthisdood** [comment]
dear god the powers gonna run out EVEN FASTER roundstart when the engitards dont set up SM?

**Comment by Pickle-Coding** [comment]
I might suggest having 4 roundstart SMES as standard for maps if load will increase. Currently roundstart demand is higher than roundstart supply, so this change might make rooms with a lot of machines run out of power quickly before engineers could set up the engine.

**Comment by ZerVexon** [comment]
Can't wait for the meta to rig up an extra two or three crates per solar array because 90% of the station population has no idea how to increase SM output without making it explode.

**Comment by nianjiilical** [comment]
> Does this mean plasma collectors take power to make power?

looking at the code it seems like things like the tesla coils and pacman inherit from obj/machinery/power which has the costs already set to 0, though thats definitely something to keep an eye out for

**Comment by Ghilker** [comment]
> > Does this mean plasma collectors take power to make power?
> 
> looking at the code it seems like things like the tesla coils and pacman inherit from obj/machinery/power which has the costs already set to 0, though thats definitely something to keep an eye out for

yes all those that produce power (ie tesla coils, pacmans) are child of power and they have their initial idle_power_usage and active_power_usage set to zero and also they have use_power set to NO_POWER_USE

File: code/modules/awaymissions/mission_code/challenge.dm [hunk 1]
```diff
@@ -34,4 +34,4 @@
 	welded = TRUE
 
 /obj/machinery/power/emitter/energycannon/RefreshParts()
-	return
+	return ..()
```
Comment (Fikou) [inline code/modules/awaymissions/mission_code/challenge.dm:37]: ?

**Comment by davethwave** [comment]
If we do this, a upgrade to the standard apc batteries may be in order. Maybe finally allow them to be upgraded via a bluespace part replacer thing.

File: code/modules/awaymissions/mission_code/challenge.dm (hunk 1 above)
Comment (Ghilker) [reply code/modules/awaymissions/mission_code/challenge.dm]: changed it

**Comment by GuiltyNeko** [comment]
From testing this locally, idle power draw is higher than what three SMES can supply, I get 747 kilowatts of draw, and medbay central runs out of power before I can finish setting up the engine. This being on meta. This is going to be a significant issue because even if nothing is turned on, the power consumption is outrageously high. Not to mention just jumping into a recharging station is a little over 3 kilowatts of draw on the APC. Not to mention that hydroponics trays with auto-grow turned on draw a grand total of 121 kilowatts. With tier four parts in all of them? 1.3 megawatts. This cannot realistically go through as it is, as roundstart power draw is impossibly high.

**Comment by Ghilker** [comment]
> From testing this locally, idle power draw is higher than what three SMES can supply, I get 747 kilowatts of draw, and medbay central runs out of power before I can finish setting up the engine. This being on meta. This is going to be a significant issue because even if nothing is turned on, the power consumption is outrageously high. Not to mention just jumping into a recharging station is a little over 3 kilowatts of draw on the APC. Not to mention that hydroponics trays with auto-grow turned on draw a grand total of 121 kilowatts. With tier four parts in all of them? 1.3 megawatts. This cannot realistically go through as it is, as roundstart power draw is impossibly high.

Yes the consumption being not achievable with the SM SMESes alone is intended, you'll need at least the basic SM and the Solars to be able to keep up with the station basic demand
This is the station draw at roundstart, notice that the supply is more than enough to keep up with it AND provide an extra when machines become active.
image

i understand that some numbers needs to be tweaked, but i think this will push engineering to improve and learn, plus we could see others try to improvise systems to increase the supply.

 
> Edit: Air alarms draw 200 when idle and 20 when active. 

yes, i'll fix it

>Edit edit: Light switches draw 200 when idle and 2000 when active, while ACTUAL lights draw a mere 80 when active and 2 when idle. 

light switches should not draw that much, or at all. Lights should draw 80 W when off (lol shouldn't consume that) and 240 W when on (which is reasonable), i'll fix those tho

>Edit edit edit: This increases the power draw in the AI chamber, up to 8.53 kilowatts of draw, or 9.54 when there's an actual AI inside the core.

AIs should draw a lot of power.

**Comment by Nabski89** [comment]
How does this play out for different maps, like Kilo(small) versus Delta(big). It seems like that might be a very frustrating Delta experience since it's the only engine that you also have to fully set up the emitters for.

**Comment by HansRobertCoooding** [comment]
another X PR to remove fun because there is too much fun in this game I guess?

**Comment by Ghilker** [comment]
ok i changed the various values and now the roundstart consumption is around this
image

so it should be fine for a long time before starting to cause issues.

stock parts got rebalanced to a lower energy multiplier so higher tier ones will require a bit less power. just a bit tho

**Comment by Ghilker** [comment]
Unless I get a feedback for a glaring balance issue or a bug or a maint asks me to change stuff this pr is basically done

**Comment by Pandarsenic** [comment]
If you know offhand, how long does each station run without losing power to major areas assuming minimal actual activation of station machinery, assuming Engineering shits the bed/isn't present at all?

**Comment by Ghilker** [comment]
> If you know offhand, how long does each station run without losing power to major areas assuming minimal actual activation of station machinery, assuming Engineering shits the bed/isn't present at all?

Aight so, every station starts with the same amount of time before any of the departments starts to lose power, that's more or less 10-15 minutes for minor rooms to start having power issues. From my observations on live servers. You get around 20 minutes before main areas lose power on delta, around 30 for kilo. I'd say that the power that you get can be comparable to the one before the change. Mostly because the main change of this pr is for active power consumption and machine tier upgrade.
Also some of the highest consumers are either not used roundstart or not built.

File: code/game/machinery/_machinery.dm [hunk 2]
```diff
@@ -988,3 +997,10 @@
 	if(isliving(user))
 		last_used_time = world.time
 		last_user_mobtype = user.type
+
+/proc/max_tierer()
+	for(var/obj/machinery/current as anything in GLOB.machines)
+		for(var/obj/item/stock_parts/part in current.component_parts)
+			part.rating = 4
+			part.energy_rating = 16
+		current.RefreshParts()
```
Comment (LemonInTheDark) [inline code/game/machinery/_machinery.dm:1006]: ????????????????

File: code/game/machinery/_machinery.dm [hunk 3]
```diff
@@ -677,8 +677,17 @@
...
+	SHOULD_CALL_PARENT(TRUE)
+	if(!component_parts || !component_parts.len)
+		return
+	var/parts_energy_rating = 0
+	for(var/obj/item/stock_parts/part in component_parts)
+		parts_energy_rating += part.energy_rating
+
+	idle_power_usage = initial(idle_power_usage) * (1 + parts_energy_rating)
+	active_power_usage = initial(active_power_usage) * (1 + parts_energy_rating)
+	update_current_power_usage()
```
Comment (LemonInTheDark) [inline code/game/machinery/_machinery.dm:690]: Shouldn't no parts still change power usage?

File: code/game/machinery/_machinery.dm [hunk 4]
```diff
@@ -107,9 +107,9 @@
 		//1 = use idle_power_usage
 		//2 = use active_power_usage
 	///the amount of static power load this machine adds to its area's power_usage list when use_power = IDLE_POWER_USE
-	var/idle_power_usage = 0
+	var/idle_power_usage = 100
 	///the amount of static power load this machine adds to its area's power_usage list when use_power = ACTIVE_POWER_USE
-	var/active_power_usage = 0
+	var/active_power_usage = 1000
```
Comment (LemonInTheDark) [inline code/game/machinery/_machinery.dm:112]: I'm not gonna go over the rest yet, cause I think this is something important to discuss

What is a power usage?
How much is "A lot"? How much is "A bit"?
How is a new "user" (Coder) supposed to figure these things out?

I would really recommend using the pattern arcane created for economy, where all values are based off a set of defines, and multipliers.

So active_power_usage `MID_DRAW * 4`, something like that.

Otherwise we will have drift, as the code changes and churns people will functionally undo your work, because the system they're working in doesn't guide them to the situation you want.
They will just do what they do currently, and pick random numbers that "look" right.

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]

File: code/game/machinery/_machinery.dm (hunk 2 above)
Comment (Ghilker) [reply code/game/machinery/_machinery.dm:1006]: Ah it was for testing, it doesn't get called anywhere, I'll remove it

File: code/game/machinery/_machinery.dm (hunk 3 above)
Comment (Ghilker) [reply code/game/machinery/_machinery.dm:690]: Well how? If you have no parts you'd still have the initial values

File: code/game/machinery/_machinery.dm (hunk 4 above)
Comment (Ghilker) [reply code/game/machinery/_machinery.dm:112]: Will do that

File: code/game/machinery/_machinery.dm (hunk 3 above)
Comment (LemonInTheDark) [reply code/game/machinery/_machinery.dm:690]: if I have parts, then lose some, I'll have my old numbers right?

Comment (Ghilker) [reply code/game/machinery/_machinery.dm:690]: Ah but currently you cannot lose parts. Unless you want me to add a check for future safeguard there isn't a way to get lower level parts

Comment (Ghilker) [reply code/game/machinery/_machinery.dm:690]: Or no part without deconstruction

Comment (LemonInTheDark) [reply code/game/machinery/_machinery.dm:690]: I'd like a safeguard.
The check also doesn't really make sense honestly, it confuses me

File: code/game/machinery/Sleeper.dm [hunk 5]
```diff
@@ -283,7 +285,6 @@
 	desc = "'Sleeper' units were once known for their healing properties, until a lengthy investigation revealed they were also dosing patients with deadly lead acetate. This appears to be one of those old 'sleeper' units repurposed as a 'Party Pod'. It’s probably not a good idea to use it."
 	icon_state = "partypod"
 	base_icon_state = "partypod"
-	idle_power_usage = 3000
```
Comment (LemonInTheDark) [inline code/game/machinery/Sleeper.dm:286]: rip in pieces

File: code/game/machinery/_machinery.dm (hunk 3 above)
Comment (Ghilker) [reply code/game/machinery/_machinery.dm:690]: i'll add the safeguard
what's wrong with the check?

File: code/__DEFINES/machines.dm [hunk 6]
```diff
@@ -27,6 +27,9 @@
 #define IDLE_POWER_USE 1
 #define ACTIVE_POWER_USE 2
 
+#define MIN_IDLE_POWER_USAGE 100
+#define MIN_ACTIVE_POWER_USAGE 1000
```
Comment (LemonInTheDark) [inline code/__DEFINES/machines.dm:31]: This isn't the most descriptive name

The main goal here is to give people setting power usage an anchor point for their choice of power usage.
It might be useful? to base this off some standard number, something people are likely to have some experience with ingame. Not totally sure what to use, but i think it'd help some

File: code/game/machinery/cell_charger.dm [hunk 7]
```diff
@@ -3,9 +3,6 @@
 	desc = "It charges power cells."
 	icon = 'icons/obj/power.dmi'
 	icon_state = "ccharger"
-	use_power = IDLE_POWER_USE
-	idle_power_usage = 5
-	active_power_usage = 60
```
Comment (LemonInTheDark) [inline code/game/machinery/cell_charger.dm:8]: Is this just because cell chargers pull on well, charge?

File: code/game/machinery/buttons.dm [hunk 8]
```diff
@@ -11,8 +11,7 @@
 	var/id = null
 	var/initialized_button = 0
 	armor = list(MELEE = 50, BULLET = 50, LASER = 50, ENERGY = 50, BOMB = 10, BIO = 100, FIRE = 90, ACID = 70)
-	use_power = IDLE_POWER_USE
-	idle_power_usage = 2
+	idle_power_usage = MIN_IDLE_POWER_USAGE * 0.02
```
Comment (LemonInTheDark) [inline code/game/machinery/buttons.dm:14]: what's goin on here?

File: code/game/machinery/dance_machine.dm [hunk 9]
```diff
@@ -464,6 +464,7 @@
 				if(!L || !L.client)
 					continue
 				L.stop_sound_channel(CHANNEL_JUKEBOX)
+		use_power(active_power_usage)
```
Comment (LemonInTheDark) [inline code/game/machinery/dance_machine.dm:467]: isn't this what the automatic usage stuff is for? or

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]

File: code/__DEFINES/machines.dm (hunk 6 above)
Comment (Ghilker) [reply code/__DEFINES/machines.dm:31]: thing is, there aren't standard numbers when it comes to power, i'm trying to standardize the amounts of 100 W for baseline idle power consumption on most machines and 1 kW for baseline active power consumption. I don't know how more descriptive or more baseline i can make this.
I think that clear and round numbers like 100 and 1000 are enough to make people understand what the baseline is and what is high amount and low amount.
For the names if you have suggestions i'm open for it

File: code/game/machinery/cell_charger.dm (hunk 7 above)
Comment (Ghilker) [reply code/game/machinery/cell_charger.dm:8]: yes, they consume power to recharge the cells faster/slower depending on parts.
Also as a note, use_power = IDLE_POWER_USE is useless to add because machinery path defaults to it already

File: code/game/machinery/buttons.dm (hunk 8 above)
Comment (Ghilker) [reply code/game/machinery/buttons.dm:14]: for use_power refer to the above comment
the rest is just 100 * 0.02 = 2 
so that buttons will draw a small 2 W

File: code/game/machinery/dance_machine.dm (hunk 9 above)
Comment (Ghilker) [reply code/game/machinery/dance_machine.dm:467]: true, i can add the procs to the right calls

**Comment by Pandarsenic** [comment]
I have... discovered another problem on Campbell.

The Syndicate Lavaland base becomes nigh unusable due to power loss if left alone for a while. It takes 8 sniper rounds just to shoot your way through the doors to the first crowbar. The Syndicate IDs can't be used to open the APCs, so you have to hack them manually to swap the free cell you have.

It's not entirely strictly unusable - I think someone who knows what they're doing can kickstart the base with the Turbine and then use that to kick off the SM shard? - but it does kind of suck for newer players who want to play around there.

**Comment by LemonInTheDark** [comment]
Needs more power

File: code/__DEFINES/machines.dm (hunk 6 above)
Comment (LemonInTheDark) [reply code/__DEFINES/machines.dm:31]: Lemme think about it.
If they are the baseline that's ok, but I really do think you want some sort of VENT_USE or something concrete people can hold in their heads. That's maybe a bad example, but you get me

Comment (Ghilker) [reply code/__DEFINES/machines.dm:31]: it's a base machine stat, i don't know how to improve the name other than having literally base machine idle consumption

File: code/modules/power/rtg.dm [hunk 10]
```diff
@@ -103,4 +104,40 @@
...
+	circuit = null
+	power_gen = 1500
+
+/obj/machinery/power/rtg/lavaland/Initialize(mapload)
+	. = ..()
+	var/turf/our_turf = get_turf(src)
+	if(!islava(our_turf))
+		power_gen = 0
+	if(!is_mining_level(z))
+		power_gen = 0
```
Comment (Fikou) [inline code/modules/power/rtg.dm:121]: this is horrible it doesnt even prevent you from moving it elsewhere, are you sure this is even laveproof?

File: code/modules/power/rtg.dm [hunk 11]
```diff
@@ -103,4 +104,40 @@
 	circuit = null
 
 /obj/machinery/power/rtg/debug/RefreshParts()
-	return
+	return ..()
```
Comment (Fikou) [inline code/modules/power/rtg.dm:107]: stop making paths that just return parent

File: code/modules/power/rtg.dm [hunk 12]
```diff
@@ -103,4 +104,40 @@
...
+		return
+	else if(default_deconstruction_crowbar(I, user = user))
+		return
+	return ..()
+
+/obj/machinery/power/rtg/old_station/default_deconstruction_crowbar(obj/item/crowbar, ignore_panel, custom_deconstruct, mob/user)
+	balloon_alert(user, "it's starting to crumble")
+	if(!do_after(user, 3 SECONDS, src))
+		return TRUE
+	balloon_alert(user, "you feel like you made a mistake")
```
Comment (Fikou) [inline code/modules/power/rtg.dm:141]: all of these balloon alerts are really bad

Comment (Ghilker) [reply code/modules/power/rtg.dm:141]: what should i put in these

File: code/modules/power/rtg.dm (hunk 10 above)
Comment (Ghilker) [reply code/modules/power/rtg.dm:121]: rtg can't be moved at all, only deconstructed. i'll add a lavaproof flag tho

File: code/modules/power/rtg.dm (hunk 11 above)
Comment (Ghilker) [reply code/modules/power/rtg.dm:107]: welp

File: code/modules/power/rtg.dm (hunk 10 above)
Comment (Ghilker) [reply code/modules/power/rtg.dm:121]: and an anchored for good measure

File: code/__DEFINES/machines.dm (hunk 6 above)
Comment (LemonInTheDark) [reply code/__DEFINES/machines.dm]: mmf

File: code/modules/power/rtg.dm (hunk 12 above)
Comment (AnturK) [reply code/modules/power/rtg.dm:151]: Just make these notice tochats really.

File: code/modules/power/rtg.dm [hunk 13]
```diff
@@ -102,5 +103,51 @@
 	power_gen = 20000
 	circuit = null
 
-/obj/machinery/power/rtg/debug/RefreshParts()
```
Comment (vincentiusvin) [inline code/modules/power/rtg.dm:105]: Why was this removed? Calling refereshparts would set the debug rtg's power to zero.

File: code/modules/power/rtg.dm (hunk 11 above)
Comment (vincentiusvin) [reply code/modules/power/rtg.dm]: bruh we need it to do nothing
